.github/workflows/tf-apply.yaml

```diff
@@ -1,21 +1,227 @@
+          }
+
+          async function typeWithRandomDelay(page, selector, text) {
+            for (const char of text) {
+              await page.type(selector, char, { delay: getRandomDelay(50, 150) });
+            }
+          }
+
+          async function waitForSelectorWithTimeout(page, selector, timeout) {
+            try {
+              await page.waitForSelector(selector, { visible: true, timeout: timeout });
+              return true;
+            } catch (error) {
+              if (error.name === 'TimeoutError') {
+                return false;
+              }
+              throw error;
+            }
+          }
+
+          (async () => {
+            const browser = await puppeteer.launch({
+              headless: true,
+              args: ['--no-sandbox', '--disable-setuid-sandbox', '--window-size=1920,1080'],
+              defaultViewport: { width: 1920, height: 1080 }
+            });
+            const page = await browser.newPage();
+            await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');
+            
+            try {
+              await page.goto('${{ env.AUTH_URL }}', { waitUntil: 'networkidle0' });
+              await page.waitForSelector('#login-username', { visible: true });
+              
+              await typeWithRandomDelay(page, '#login-username', process.env.SPOTIFY_USERNAME);
+              await delay(getRandomDelay(500, 1500));
+              await typeWithRandomDelay(page, '#login-password', process.env.SPOTIFY_PASSWORD);
+              
+              await delay(getRandomDelay(1000, 2000));
+              await page.click('#login-button');
+              
+              // Wait for either the authorization button or successful authorization
+              const authAcceptSelector = '#auth-accept';
+              const authSuccessTimeout = 20000; // 20 seconds
+              
+              const isAuthAcceptPresent = await waitForSelectorWithTimeout(page, authAcceptSelector, authSuccessTimeout);
+              
+              if (isAuthAcceptPresent) {
+                console.log('Permission request detected. Accepting...');
+                await page.click(authAcceptSelector);
+                await page.waitForNavigation({ waitUntil: 'networkidle0', timeout: 30000 });
+              }
+              
+              const content = await page.content();
+              if (content.includes('Authorization successful') || content.includes('callback?code=')) {
+                console.log('Spotify authorization successful');
+              } else {
+                throw new Error('Authorization not successful');
+              }
+            } catch (error) {
+              console.error('Error during Spotify login or authorization:', error);
+              process.exit(1);
+            } finally {
+              await browser.close();
+            }
+          })();
+          EOF
```

Comment:
can we have this logic outside of the yaml file, maybe within a file/directory dedicated to authentication and just call that here?

Reply:
i tried it yesterday and gave up 💀 
it seems trivial to do but when i did...auth didn't work (the puppeteer part)
then i reverted the changes to what was working earlier, but auth still failed. the literal same code is not working now :( 
i can only believe it's spotify blocking login from unknown ip's

Reply:
as for the tfstate files...
they are required to be version controlled either here or on something like an s3 bucket (preferable)
without that, the github actions workflow creates a new playlist on every push